src/lib.rs

```diff
@@ -0,0 +1,277 @@
+        let mut e: Scalar = Scalar::random(&mut *rng);
+
+        // Compute the scalar inverse: (x + e)^-1.
+        let mut d: Scalar = *x + e; // d = x + e
+        // We loop, re-sampling 'e' until (x + e) is non-zero and we get an inverse.
+        let d_inv: Scalar = loop {
+            // Attempt to compute the inverse.
+            // ff::Field::invert returns subtle::CtOption<Self>.
+            let inv = d.invert();
+            if bool::from(inv.is_some()) {
+                // Success! The inverse exists.
+                break inv.unwrap();
+            } else {
+                // (x + e) was zero. Sample a new 'e' and the loop will retry.
+                e = Scalar::random(&mut *rng);
+                d = *x + e;
+            }
+        };
+
+        // Compute the multi-scalar multiplication with the generators and messages:
+        // `B = G + m1*H1 + ... + ml*Hl`
+        // TODO do we have some form of multi-exp here?
+        let b: G1Projective = bbs_params.g_generator
+            + messages
+                .iter()
+                .zip(bbs_params.h_generators.iter())
+                .map(|(m, h)| (*h) * (*m))
+                .sum::<G1Projective>();
+
+        // Compute the final signature component `A = d_inv * B`
+        let a: G1Projective = b * d_inv;
+
+        Ok(Self { a, e })
+    }
+
+    /// Verifies a BBS MAC by recomputing B and checking that `A == (x + e)^{-1} * B`.
+    ///
+    /// # Parameters
+    /// - x: the secret key scalar
+    /// - messages: the messages to be verified
+    /// - mac: the MAC to be verified
+    /// - bbs_params: the system parameters
+    ///
+    /// # Returns
+    /// - `Ok(())`: the MAC is valid
+    /// - `Err(BbsMacError::VerificationInverseFailed)`: could not invert (x + e)
+    /// - `Err(BbsMacError::InputLengthMismatch)`: number of messages does not match number of generators
+    /// - `Err(BbsMacError::VerificationFailed)`: computed value does not match the provided MAC
+    ///
+    /// # Example
+    /// ```rust
+    /// use bbs_mac_vs_sig::{BbsParams, BbsMac};
+    /// use bls12_381::Scalar;
+    /// use ff::Field;
+    ///
+    /// let mut rng = rand::thread_rng();
+    /// let num_messages = 5;
+    /// let params = BbsParams::new(num_messages, &mut rng);
+    /// let x = Scalar::random(&mut rng);
+    ///
+    /// let mut messages = Vec::with_capacity(num_messages);
+    /// for _ in 0..num_messages {
+    ///     messages.push(Scalar::random(&mut rng));
+    /// }
+    ///
+    /// let mac = BbsMac::sign(&x, &messages, &params, &mut rng).unwrap();
+    /// BbsMac::verify(&x, &messages, &mac, &params).unwrap();
+    /// ```
+    pub fn verify(
+        x: &Scalar,
+        messages: &[Scalar],
+        mac: &BbsMac,
+        bbs_params: &BbsParams<G1Projective>,
+    ) -> Result<(), BbsMacError> {
+        let inv = (x + mac.e).invert();
+        if bool::from(inv.is_none()) {
+            return Err(BbsMacError::VerificationInverseFailed);
+        }
```

Comment:
Cleaner control flow
```suggestion
        let Some(inv) = (x + mac.e).invert() else {
            return Err(BbsMacError::VerificationInverseFailed);
        }
```

Reply:
added